java/java-analysis-impl/src/com/intellij/psi/impl/source/resolve/reference/impl/JavaReflectionReferenceUtil.java

```diff
@@ -468,6 +476,39 @@ public static boolean isCallToMethod(@NotNull PsiMethodCallExpression methodCall
     return null;
   }
 
+  /**
+   * Tries to unwrap array or {@link List} and find its components
+   *
+   * @param maybeArrayOrList an array or {@link List} to unwrap
+   * @return list of unwrapped array or list components, some or all of them could be null if unknown (but the length is known);
+   * returns null if nothing is known.
+   */
+  @Nullable
+  public static List<PsiExpression> getArrayOrListComponents(@Nullable PsiExpression maybeArrayOrList) {
+    if (maybeArrayOrList == null) {
+      return null;
+    }
+    final List<PsiExpression> varargs = getVarargs(maybeArrayOrList);
+    if (varargs != null) {
+      return varargs;
+    }
+    if (maybeArrayOrList instanceof PsiMethodCallExpression) {
+      final PsiMethodCallExpression callExpression = (PsiMethodCallExpression)maybeArrayOrList;
+      if (isCallToMethod(callExpression, JAVA_UTIL_LIST, OF)
+          || isCallToMethod(callExpression, JAVA_UTIL_ARRAYS, AS_LIST)) {
+        final PsiExpression[] expressions = callExpression.getArgumentList().getExpressions();
+        if (expressions.length == 1) {
```

Comment:
I replaced the check with `MethodCallUtils#isVarArgCall`

Now, the inspection does not give a warning for the following code:

```java
class Test2 {
    void bar() {
    }
}

public class Main2 {
    public static void main(String[] args) throws Throwable {
        MethodHandles.Lookup lookup = MethodHandles.lookup();
        MethodHandle methodHandle = lookup
                .findVirtual(Test2.class, "bar", MethodType.methodType(Test2.class, List.of(new Class<?>[]{int.class, long.class})));
        methodHandle.invoke(new Test2());
    }
}
```

Maybe, we can not support this? Because:

1. To me it seems too artificial to support
2. Arrays could be mutated if they are created previously and passed as a variable to the call

---

> 2. Arrays could be mutated if they are created previously and passed as a variable to the call

A similar issue exists for reflection inspection. Consider the following code:

```java
class Test3 {
    void bar(int unused) {
        System.out.println("bar(int)");
    }
}

public class Main3 {
    public static void main(String[] args) throws Throwable {
        Class<?> clazz = Test3.class;
        Class<?>[] parameterTypes = {long.class};
        parameterTypes[0] = int.class;
        Method method = clazz.getDeclaredMethod("bar", parameterTypes);
        //                                             ^^ Cannot resolve method 'bar' with specified argument types
        method.invoke(new Test3(), 10);
    }
}
```

IDEA produces a warning, thought the `bar` method exists

Should it be fixed too? Maybe, if this MR would merged, I can submit another one?